src/librustc_mir/transform/generator.rs

```diff
@@ -539,29 +661,37 @@ fn compute_layout<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     }
 
     let dummy_local = LocalDecl::new_internal(tcx.mk_unit(), mir.span);
-
-    // Gather live locals and their indices replacing values in mir.local_decls with a dummy
-    // to avoid changing local indices
-    let live_decls = live_locals.iter().map(|local| {
-        let var = mem::replace(&mut mir.local_decls[local], dummy_local.clone());
-        (local, var)
+    let mut remap = FxHashMap::default();
+
+    const INITIAL_VARIANTS: usize = 3;
+    let prefix_variant = VariantIdx::new(INITIAL_VARIANTS);
+    let prefix_fields = prefix_locals.iter().enumerate().map(|(idx, local)| {
+        // Gather LocalDecls of live locals replacing values in mir.local_decls
+        // with a dummy to avoid changing local indices.
+        let field = mem::replace(&mut mir.local_decls[local], dummy_local.clone());
+        remap.insert(local, (field.ty, prefix_variant, idx));
+        field
+    })
+    .collect::<Vec<_>>();
+
+    let state_variants = suspending_blocks.iter().enumerate().map(|(variant, block)| {
+        let variant_index = VariantIdx::new(INITIAL_VARIANTS + variant);
+        let no_assignments = &liveness::LiveVarSet::new_empty(0);
+        let variant_locals = variant_assignments.get(&block)
+            .unwrap_or(no_assignments)
+            .iter().enumerate().map(|(idx, local)| {
+                // Gather LocalDecls of live locals replacing values in
+                // mir.local_decls with a dummy to avoid changing local indices.
+                let field = mem::replace(&mut mir.local_decls[local], dummy_local.clone());
+                // These fields follow the prefix fields.
+                remap.insert(local, (field.ty, variant_index, prefix_fields.len() + idx));
+                field
+            });
+        prefix_fields.iter().cloned().chain(variant_locals).collect::<Vec<_>>()
```

Comment:
Are you relying on the `prefix_fields` having the same offsets in all the variants here?

I did not see any code which ensures this happen.

Reply:
Yeah you're right, I need to rethink what information we need in the `GeneratorLayout` struct to do this properly.

Reply:
In general I feel like representing generators as enum is not a great fit.

I had imagined the generator transformation treating it as a struct (without having variants which complicates the transformation). The transformation would produce information about which locals may overlap. The layout code (which has access to the final types) would then use that information to potentially overlap fields. This structure would allow more powerful layout optimization than the one you implemented here.

Reply:
We had discussed something like this very early on. Something like that can work, in theory, but there are a few problems to be solved...

1. We can't describe structs with overlapping fields in debuginfo
2. Ideally, we would have a way to describe invariants on the generator fields, to be checked with a UB sanitizer using miri. This requires that (after the generator returns) we know the relationship between the state field of the generator and which fields are initialized.

Using variant layouts solves both of these problems. In fact, in my mind we are effectively doing what you described, but using variant layouts as an "implementation detail."

It does present the problem you note here, where we have to encode our expectations in `GeneratorLayout` about which fields are the same between variants. This relationship is really only important during the transform stage and in layout computation; everything downstream of that doesn't really care. (Debuginfo also cares, but that's already solved; we just need to continue to save off the `LocalDecl`s for later.)

Another option using variant layouts (that I actually started doing at one point) is to put the fields that never move in the "prefix" of our generator, after the upvars. I would say this provides less overall flexibility, in case we wanted to have, for example, a field that is only live for Suspends 2-4.

> This structure would allow more powerful layout optimization than the one you implemented here.

Can you be more specific about how this would allow a more powerful optimization? Or maybe we're saying the same thing: we need to give the layout computation a more complete picture of what's going on. I'm just saying that after we can continue to use variant layouts after that point.

cc @eddyb

Reply:
...so, to be more specific, I propose that we make `GeneratorLayout` look something like this:

```rust
struct GeneratorLayout<'tcx> {
  field_tys: Vec<Ty<'tcx>>,
  variant_fields: Vec<Vec<u32>>,  // where each element is an index into `field_tys`
}
```

and put debuginfo somewhere else. Thus we'll erase the information about where the fields came from, but preserve the information about which fields are the same across variants. Layout computation can use this to ensure that a given field has the same offset across all variants.

This means our MIR transform will be relying on layout computation to enforce this contract, or we will generate invalid code. I think this is fine if we document it well. Those are the only two pieces of code that need to care about this.

Reply:
We can probably use `Variants::Multiple` to represent the layout of generators, but we can't use the same code to compute the layout as enums as we have addition restrictions. I also don't think it's a good idea to "pollute" the enum code with generator specific concerns.

I'm not sure your proposed `GeneratorLayout` is sufficient to express the constraints we need.
It doesn't handle the case where have two locals `a` and `b` which are in 2 different variants/states, but their livetimes overlap in between.
```rust
let a = A;
yield; // Only `a` is live here
let b = B;
drop(a);
yield; // Only `b` is live here
drop(b);
```

Reply:
> I'm not sure your proposed GeneratorLayout is sufficient to express the constraints we need.

The way the code is now, those constraints are all dealt with during the MIR transform stage, when the `GeneratorLayout` is computed. In your example the code would decide that `a` and `b` cannot overlap, and thus stick at least one of them in every variant, to prevent them from overlapping.

I'm not saying this is the best factoring, or even a good one. It seemed like the most convenient (and maybe efficient) place to do this analysis, but tbh I didn't really consider other places.

In theory, we could put all of the info about which locals have overlapping storage in our `GeneratorLayout` and then do the calculation as part of the layout. The problem is, this information is O(N^2) in the number of locals. _(More precisely, we need to keep around N^2/2 bits to store it, where N is the number of locals. We can renumber just the locals we store to keep it small)._ Maybe this isn't really a problem in practice, but it's something to think about.

Reply:
> The way the code is now, those constraints are all dealt with during the MIR transform stage, when the `GeneratorLayout` is computed. In your example the code would decide that `a` and `b` cannot overlap, and thus stick at least one of them in every variant, to prevent them from overlapping.
> 
> I'm not saying this is the best factoring, or even a good one. It seemed like the most convenient (and maybe efficient) place to do this analysis, but tbh I didn't really consider other places.

We can probably start out with this trick to resolve constraints and transition to giving the layout code more information later.


> I think I'd personally prefer to have the logic that produces it in the same place as the logic that consumes it.

The generator transformation operates pre-monomorphization so we can't do final layout there so we need an abstract representation to pass on to the layout code operating post-monomorphization. Ideally one flexible enough for the layout code to produce an ideal result (even though computing it would be NP-hard).

Reply:
> The generator transformation operates pre-monomorphization so we can't do final layout there so we need an abstract representation to pass on to the layout code operating post-monomorphization.

:+1: 

> Ideally one flexible enough for the layout code to produce an ideal result (even though computing it would be NP-hard).

Yeah, that's the one I'm saying probably needs to be O(N^2) in the number of saved locals. I really don't know if that's a problem. Maybe it's better to start with this so we don't have to regress performance later? =)

cc @eddyb @nikomatsakis , any thoughts?

Reply:
I think `N` is practically bounded because of two reasons:

- Users avoid creating large generators
- Inlining will stop once generator reaches a certain size (though we don't do any inlining currently)